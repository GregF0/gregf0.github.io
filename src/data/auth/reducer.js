export const initialState = {
  uid: null,
  name: null,
  photo: null,
  creationTime: null,
}

export const reducer = (state = initialState, action) => {
  switch (action.type) {
    case 'SET_USER_INFO':
      return {
        ...state,
        ...action.userInfo,
      }

    case 'SIGN_OUT':
      return {
        ...initialState,
      }
    default:
      return state
  }
}
