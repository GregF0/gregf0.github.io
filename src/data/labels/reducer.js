export const initialState = {
  fullscreenDialog: false,
  desktopDialog: false,
  deleteAlert: false,
  menuOpened: false,
  formValue: null,
  labelEditting: null,
  labelSelected: null,
  data: [],
}

export const reducer = (state = initialState, action) => {
  switch (action.type) {
    case 'SET_FULLSCREEN_DIALOG':
      return {
        ...state,
        fullscreenDialog: action.fullscreenDialog,
      }
    case 'SET_DESKTOP_DIALOG':
      return {
        ...state,
        desktopDialog: action.desktopDialog,
      }
    case 'SET_DELETE_ALERT':
      return {
        ...state,
        deleteAlert: action.deleteAlert,
      }
    case 'SET_MENU_OPENED':
      return {
        ...state,
        menuOpened: action.menuOpened,
      }
    case 'SET_LABEL_EDITTING':
      return {
        ...state,
        labelEditting: action.labelEditting,
      }
    case 'SET_FORM_VALUE':
      return {
        ...state,
        formValue: action.formValue,
      }
    case 'ADD_LABEL':
      return {
        ...state,
        data: [...state.data, action.label],
        labelSelected: state.labelSelected || action.label,
      }
    case 'EDIT_LABEL':
      const updatedLabels = state.data.map((label) => {
        if (label.id === action.id) {
          return {
            ...label,
            ...action.updates,
          }
        } else {
          return label
        }
      })
      return {
        ...state,
        data: updatedLabels,
        labelSelected:
          state.labelSelected && state.labelSelected.id === action.id
            ? { ...state.labelSelected, ...action.updates }
            : state.labelSelected,
      }
    case 'DELETE_LABEL':
      return {
        ...state,
        data: state.data.filter(({ id }) => id !== action.id),
        labelSelected:
          state.labelSelected && state.labelSelected.id === action.id
            ? null
            : state.labelSelected,
      }
    case 'SET_LABELS':
      return {
        ...state,
        data: action.labels,
      }
    case 'SET_LABEL_SELECTED':
      return {
        ...state,
        labelSelected: action.labelSelected,
      }
    default:
      return state
  }
}
