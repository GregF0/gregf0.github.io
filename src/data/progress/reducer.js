const initialState = true

export const reducer = (state = initialState, action) => {
  switch (action.type) {
    case 'SET_PROGRESS_VISIBILITY':
      return action.visibility
    default:
      return state
  }
}
