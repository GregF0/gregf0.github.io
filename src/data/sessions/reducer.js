const initialState = []

export const reducer = (state = initialState, action) => {
  switch (action.type) {
    case 'SET_SESSIONS':
      return [...action.sessions]
    case 'ADD_SESSION':
      return [...state, action.session]
    default:
      return state
  }
}
