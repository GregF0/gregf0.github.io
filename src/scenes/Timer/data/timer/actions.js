export const setStatus = (status) => ({
  type: 'SET_STATUS',
  status,
})

export const setProgress = (progress) => ({
  type: 'SET_PROGRESS',
  progress,
})

export const setTimeLeft = (timeLeft) => ({
  type: 'SET_TIME_LEFT',
  timeLeft,
})

export const saveInterval = (interval) => ({
  type: 'SAVE_INTERVAL',
  interval,
})

export const pauseTimer = () => ({
  type: 'PAUSE_TIMER',
})

export const resetTimer = ({ duration, showTimerInTitle }) => ({
  type: 'RESET_TIMER',
  duration,
  showTimerInTitle,
})

export const setNextTimer = (settings) => ({
  type: 'SET_NEXT_TIMER',
  settings,
})

export const setSaveSessionAlert = (saveSessionAlert) => ({
  type: 'SET_SAVE_SESSION_ALERT',
  saveSessionAlert,
})
