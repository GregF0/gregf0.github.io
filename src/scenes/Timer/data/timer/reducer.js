export const STATUSES = {
  onHold: 'ON_HOLD',
  running: 'RUNNING',
  paused: 'PAUSED',
}

export const TYPES = {
  work: 'WORK',
  shortBreak: 'SHORT_BREAK',
  longBreak: 'LONG_BREAK',
}

export const initialState = {
  status: STATUSES.onHold,
  type: TYPES.work,
  progress: 100,
  timeLeft: null,
  currentRound: 1,
  interval: null,
  saveSessionAlert: false,
}

export const reducer = (state = initialState, action) => {
  switch (action.type) {
    case 'SET_STATUS':
      return {
        ...state,
        status: action.status,
      }
    case 'SET_PROGRESS':
      return {
        ...state,
        progress: action.progress,
      }
    case 'SET_TIME_LEFT':
      const { minutes, seconds } = action.timeLeft

      return {
        ...state,
        timeLeft: {
          minutes: minutes,
          seconds: isNaN(seconds) ? 0 : seconds,
        },
      }

    case 'SAVE_INTERVAL':
      return {
        ...state,
        interval: action.interval,
      }
    case 'PAUSE_TIMER':
      clearInterval(state.interval)

      return {
        ...state,
        status: STATUSES.paused,
        interval: null,
      }
    case 'RESET_TIMER':
      clearInterval(state.interval)
      const timeLeft = { minutes: action.duration, seconds: 0 }

      if (action.showTimerInTitle) {
        setTitle(state.type, timeLeft)
      }

      return {
        ...state,
        status: STATUSES.onHold,
        interval: null,
        timeLeft,
        progress: 100,
      }
    case 'SET_NEXT_TIMER':
      clearInterval(state.interval)

      const {
        workDuration,
        shortBreakDuration,
        longBreakDuration,
        rounds,
        showTimerInTitle,
      } = action.settings

      let newType
      let newTimeLeft
      let newCurrentRound = state.currentRound

      if (state.currentRound < rounds) {
        if (state.type === TYPES.work) {
          newType = TYPES.shortBreak
          newTimeLeft = { minutes: shortBreakDuration, seconds: 0 }
        } else {
          newType = TYPES.work
          newTimeLeft = { minutes: workDuration, seconds: 0 }
          newCurrentRound = state.currentRound + 1
        }
      } else {
        if (state.type === TYPES.work) {
          newType = TYPES.longBreak
          newTimeLeft = { minutes: longBreakDuration, seconds: 0 }
        } else {
          newType = TYPES.work
          newTimeLeft = { minutes: workDuration, seconds: 0 }
          newCurrentRound = 1
        }
      }

      if (showTimerInTitle) {
        setTitle(newType, newTimeLeft)
      }

      return {
        ...state,
        status: STATUSES.onHold,
        progress: 100,
        interval: null,
        type: newType,
        timeLeft: newTimeLeft,
        currentRound: newCurrentRound,
      }

    case 'SET_SAVE_SESSION_ALERT':
      return {
        ...state,
        saveSessionAlert: action.saveSessionAlert,
      }
    default:
      return state
  }
}

export const setTitle = (type, timeLeft) => {
  let emoji

  if (type === TYPES.work) {
    emoji = 'üë®‚Äçüíª'
  } else {
    emoji = '‚òïÔ∏è'
  }

  document.title = `${
    timeLeft.minutes < 10 ? '0' + timeLeft.minutes : timeLeft.minutes
  }:${
    timeLeft.seconds < 10 ? '0' + timeLeft.seconds : timeLeft.seconds
  } ${emoji} | Pomodor`
}
