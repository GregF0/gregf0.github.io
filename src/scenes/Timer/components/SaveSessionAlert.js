import React from 'react'
import { useDispatch, useSelector } from 'react-redux'
import styled from 'styled-components'
import { useTheme } from '@material-ui/core'
import Dialog from '@material-ui/core/Dialog'
import DialogTitle from '@material-ui/core/DialogTitle'
import DialogContent from '@material-ui/core/DialogContent'
import DialogContentText from '@material-ui/core/DialogContentText'
import DialogActions from '@material-ui/core/DialogActions'
import Button from '@material-ui/core/Button'
import { setSaveSessionAlert } from '../data/timer/actions'
import { startAddSession } from '../../../data/sessions/actions'

const Time = styled.span`
  color: ${({ color }) => color};
`

export const SaveSessionAlert = ({ time }) => {
  const { saveSessionAlert } = useSelector((state) => state.timer)
  const label = useSelector((state) => state.labels.labelSelected)

  const theme = useTheme()

  const dispatch = useDispatch()

  const closeAlert = () => {
    dispatch(setSaveSessionAlert(false))
  }

  const saveSession = () => {
    dispatch(
      startAddSession({
        duration: time,
        label: label ? label.id : null,
        createdAt: Date.now(),
      })
    )
    closeAlert()
  }

  return (
    <Dialog
      open={saveSessionAlert}
      onClose={closeAlert}
      aria-labelledby="save-session-alert-title"
      aria-describedby="save-session-alert-description"
    >
      <DialogTitle id="save-session-alert-title">Save session?</DialogTitle>
      <DialogContent>
        <DialogContentText id="save-session-alert-description">
          Do you want to save{' '}
          <Time color={theme.palette.secondary.main}>
            {time.minutes < 10 ? '0' + time.minutes : time.minutes}:
            {time.seconds < 10 ? '0' + time.seconds : time.seconds}
          </Time>{' '}
          to the statistics?
        </DialogContentText>
      </DialogContent>
      <DialogActions>
        <Button onClick={closeAlert}>Don't save</Button>
        <Button onClick={saveSession} color="primary">
          Save
        </Button>
      </DialogActions>
    </Dialog>
  )
}
