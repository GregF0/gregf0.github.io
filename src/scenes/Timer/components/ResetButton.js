import React, { useState } from 'react'
import styled from 'styled-components'
import IconButton from '@material-ui/core/IconButton'
import ReplayIcon from '@material-ui/icons/Replay'
import { useDispatch, useSelector } from 'react-redux'
import { resetTimer, setSaveSessionAlert } from '../data/timer/actions'
import { TYPES, STATUSES } from '../data/timer/reducer'
import { SaveSessionAlert } from './SaveSessionAlert'

export const ResetButton = () => {
  const { type, timeLeft, status } = useSelector((state) => state.timer)
  const {
    workDuration,
    shortBreakDuration,
    longBreakDuration,
    showTimerInTitle,
  } = useSelector((state) => state.settings)

  const [timeToSave, setTimeToSave] = useState({ minutes: 0, seconds: 0 })

  const dispatch = useDispatch()

  const handleClick = () => {
    if (type === TYPES.work) {
      const secondsTotal = workDuration * 60
      const secondsLeft = timeLeft.minutes * 60 + timeLeft.seconds
      const secondsDiff = secondsTotal - secondsLeft

      const minutesToSave = Math.floor(secondsDiff / 60)
      const secondsToSave = secondsDiff % 60

      if (minutesToSave) {
        setTimeToSave({
          minutes: minutesToSave,
          seconds: secondsToSave,
        })

        dispatch(setSaveSessionAlert(true))
      }
    }

    let duration

    switch (type) {
      case TYPES.work:
        duration = workDuration
        break
      case TYPES.shortBreak:
        duration = shortBreakDuration
        break
      case TYPES.longBreak:
        duration = longBreakDuration
        break
      default:
        break
    }
    dispatch(resetTimer({ duration, showTimerInTitle }))
  }

  return (
    <>
      <ActionIcon
        disabled={status === STATUSES.onHold}
        aria-label="Reset timer"
        onClick={handleClick}
        size="small"
      >
        <ReplayIcon />
      </ActionIcon>
      <SaveSessionAlert time={timeToSave} />
    </>
  )
}

export const ActionIcon = styled(IconButton)`
  border: 1px solid #bababa;
`
