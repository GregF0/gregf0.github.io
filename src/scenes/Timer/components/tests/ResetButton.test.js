import React from 'react'
import * as redux from 'react-redux'
import { createShallow } from '@material-ui/core/test-utils'
import { ResetButton, ActionIcon } from '../ResetButton'
import { TYPES, STATUSES } from '../../data/timer/reducer'
import { SaveSessionAlert } from '../SaveSessionAlert'

describe('<ResetButton />', () => {
  const shallow = createShallow()
  const createWrapper = () => {
    return shallow(<ResetButton />)
  }

  const dispatchMocked = jest.fn()
  jest.spyOn(redux, 'useDispatch').mockImplementation(() => dispatchMocked)

  const createStore = (
    status = STATUSES.onHold,
    type = TYPES.workDuration,
    timeLeft = { minutes: 25, seconds: 0 }
  ) => {
    const store = {
      timer: { status, type, timeLeft, saveSessionAlert: false },
      settings: {
        workDuration: 25,
        shortBreakDuration: 5,
        longBreakDuration: 20,
        showTimerInTitle: true,
      },
      labels: {
        labelSelected: null,
      },
    }

    jest
      .spyOn(redux, 'useSelector')
      .mockImplementation((callback) => callback(store))

    return store
  }

  beforeEach(() => {
    jest.clearAllMocks()
  })

  test('should render <ResetButton /> correctly', () => {
    createStore()
    expect(createWrapper()).toMatchSnapshot()
  })

  test('should be disabled if the timer is on hold', () => {
    createStore()
    expect(createWrapper().find(ActionIcon).prop('disabled')).toBe(true)
  })

  test('should be enabled if the timer is running', () => {
    createStore(STATUSES.running, TYPES.work, {
      minutes: 22,
      seconds: 10,
    })

    expect(createWrapper().find(ActionIcon).prop('disabled')).toBe(false)
  })

  test('should be enabled if the timer is paused', () => {
    createStore(STATUSES.paused, TYPES.work, {
      minutes: 22,
      seconds: 10,
    })

    expect(createWrapper().find(ActionIcon).prop('disabled')).toBe(false)
  })

  test('should display save session alert with the remaining time if the work timer has been running for 1 or more minutes', () => {
    createStore(STATUSES.running, TYPES.work, {
      minutes: 22,
      seconds: 10,
    })

    const wrapper = createWrapper()

    wrapper.find(ActionIcon).simulate('click')

    expect(wrapper.find(SaveSessionAlert).prop('time')).toEqual({
      minutes: 2,
      seconds: 50,
    })
  })

  test('should not display save session alert if the timer has been running for less than a minute', () => {
    createStore(STATUSES.running, TYPES.work, {
      minutes: 24,
      seconds: 10,
    })

    const wrapper = createWrapper()

    wrapper.find(ActionIcon).simulate('click')

    expect(wrapper.find(SaveSessionAlert).prop('time')).toEqual({
      minutes: 0,
      seconds: 0,
    })
  })

  test('should not display save session alert if the timer type is short break', () => {
    createStore(STATUSES.running, TYPES.shortBreak, {
      minutes: 3,
      seconds: 10,
    })

    const wrapper = createWrapper()

    wrapper.find(ActionIcon).simulate('click')

    expect(wrapper.find(SaveSessionAlert).prop('time')).toEqual({
      minutes: 0,
      seconds: 0,
    })
  })

  test('should not display save session alert if the timer type is long break', () => {
    createStore(STATUSES.running, TYPES.longBreak, {
      minutes: 11,
      seconds: 10,
    })

    const wrapper = createWrapper()

    wrapper.find(ActionIcon).simulate('click')

    expect(wrapper.find(SaveSessionAlert).prop('time')).toEqual({
      minutes: 0,
      seconds: 0,
    })
  })
})
