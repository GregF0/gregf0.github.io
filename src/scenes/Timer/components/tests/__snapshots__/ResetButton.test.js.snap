// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ResetButton /> should render <ResetButton /> correctly 1`] = `
<Fragment>
  <Styled(WithStyles(ForwardRef(IconButton)))
    aria-label="Reset timer"
    disabled={true}
    onClick={[Function]}
    size="small"
  >
    <ReplayIcon />
  </Styled(WithStyles(ForwardRef(IconButton)))>
  <SaveSessionAlert
    time={
      Object {
        "minutes": 0,
        "seconds": 0,
      }
    }
  />
</Fragment>
`;
